app/controllers/api/databases.php

```diff
@@ -1879,42 +1889,72 @@ function createAttribute(string $databaseId, string $collectionId, Document $att
             }
         }
 
-        // Check collection permissions when enforced
-        if ($collection->getAttribute('permission') === 'collection') {
-            $validator = new Authorization('write');
-            if (!$validator->isValid($collection->getWrite())) {
-                throw new Exception(Exception::USER_UNAUTHORIZED);
-            }
+        $validator = new Authorization('create');
+        if (!$validator->isValid($collection->getCreate())) {
+            throw new Exception(Exception::USER_UNAUTHORIZED);
         }
 
-        $data['$collection'] = $collection->getId(); // Adding this param to make API easier for developers
-        $data['$id'] = $documentId == 'unique()' ? $dbForProject->getId() : $documentId;
-        $data['$read'] = (is_null($read) && !$user->isEmpty()) ? ['user:' . $user->getId()] : $read ?? []; //  By default set read permissions for user
-        $data['$write'] = (is_null($write) && !$user->isEmpty()) ? ['user:' . $user->getId()] : $write ?? []; //  By default set write permissions for user
-
-        // Users can only add their roles to documents, API keys and Admin users can add any
-        $roles = Authorization::getRoles();
+        /**
+         * Map aggregate permissions into the multiple permissions they represent,
+         * accounting for the resource type given that some types not allowed specific permissions.
+         */
+        $permissions = PermissionsProcessor::aggregate($permissions, 'document');
 
-        if (!Auth::isAppUser($roles) && !Auth::isPrivilegedUser($roles)) {
-            foreach ($data['$read'] as $read) {
-                if (!Authorization::isRole($read)) {
-                    throw new Exception(Exception::USER_UNAUTHORIZED, 'Read permissions must be one of: (' . \implode(', ', $roles) . ')');
+        /**
+         * Add permissions for current the user for any missing types
+         * from the allowed permissions for this resource type.
+         */
+        $allowedPermissions = \array_filter(
```

Comment:
The logic here should be much simpler. If the permissions array is empty and we have document security enabled, we give the current user read+write permissions. Same should apply for files.

We don't try to provide fallbacks for missing permissions.

Reply:
Happy to change as suggested, just to be clear though, that will differ from the existing code, where we do provide fallbacks: 

https://github.com/appwrite/appwrite/blob/master/app/controllers/api/databases.php#L1874-L1875